packages/network-controller/src/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,2673 @@
+            const request = {
+              method,
+              // Note that `blockParam` is `0x100` here
+              params: buildMockParams({ blockParamIndex, blockParam }),
+            };
+            const mockResult = emptyValue;
+
+            await withMockedCommunications({ providerType }, async (comms) => {
+              // The first time a block-cacheable request is made, the latest block
+              // number is retrieved through the block tracker first.
+              comms.mockNextBlockTrackerRequest({ blockNumber: '0x42' });
+              comms.mockRpcCall({
+                request: buildRequestWithReplacedBlockParam(
+                  request,
+                  blockParamIndex,
+                  '0x100',
+                ),
+                response: { result: mockResult },
+              });
+
+              const result = await withNetworkClient(
+                { providerType },
+                ({ makeRpcCall }) => makeRpcCall(request),
+              );
+
+              expect(result).toStrictEqual(mockResult);
+            });
+          });
+
+          it(`does not reuse the result of a previous request if it was "${emptyValue}"`, async () => {
+            const requests = [
+              {
+                method,
+                // Note that `blockParam` is `0x100` here
+                params: buildMockParams({ blockParamIndex, blockParam }),
+              },
+              {
+                method,
+                // Note that `blockParam` is `0x100` here
+                params: buildMockParams({ blockParamIndex, blockParam }),
+              },
+            ];
+            const mockResults = [emptyValue, { blockHash: '0x100' }];
+
+            await withMockedCommunications({ providerType }, async (comms) => {
+              // The first time a block-cacheable request is made, the latest block
+              // number is retrieved through the block tracker first.
+              comms.mockNextBlockTrackerRequest({ blockNumber: '0x42' });
+              comms.mockRpcCall({
+                request: buildRequestWithReplacedBlockParam(
+                  requests[0],
+                  blockParamIndex,
+                  '0x100',
+                ),
+                response: { result: mockResults[0] },
+              });
+              comms.mockRpcCall({
+                request: buildRequestWithReplacedBlockParam(
+                  requests[1],
+                  blockParamIndex,
+                  '0x100',
+                ),
+                response: { result: mockResults[1] },
+              });
+
+              const results = await withNetworkClient(
+                { providerType },
+                ({ makeRpcCallsInSeries }) => makeRpcCallsInSeries(requests),
+              );
+
+              expect(results).toStrictEqual(mockResults);
+            });
+          });
+        }
+      });
+    }
+  });
+
+  if (method !== 'eth_getTransactionCount') {
```

Comment:
In the PR description you said that `eth_getTransactionCount` breaks on tests dealing with a block tag of `pending`, but did you figure out why this was? No worries if not, I can take a closer look at this tomorrow.

Reply:
I can't remember exactly what it was but I think it was related to the nonce tracker

Reply:
@mcmire heres the code that makes it not work:
https://github.com/MetaMask/web3-provider-engine/blob/main/subproviders/hooked-wallet.js#L637